spreadsheets.js

```diff
@@ -206,38 +214,93 @@ function find_pairing(service_account_auth, spreadsheet_key, white, black, callb
-                if (potential_rows.length > 1) {
-                    return callback("Unable to find pairing. More than one row was returned!");
-                } else if(potential_rows.length < 1) {
-                    return callback("Unable to find pairing. No rows were returned!");
+            options = _.clone(options);
+            var defaults = {
+                'min-row': 1,
+                'max-row': 100,
+                'min-col': 1,
+                'max-col': 8,
+                'return-empty': true
+            }
+            options = _.defaults(options, defaults);
+            pairings_sheet.getCells(
+                options,
+                function(err, cells) {
+                    if (err) { return callback(err, cells); }
+
+                    // Take the single list and turn it into rows
+                    var rows = [];
+                    cells.forEach(function(cell) {
+                        // Create the empty row if needed.
+                        if (!(cell.row in rows)) {
+                            rows[cell.row] = [];
+                        }
+                        rows[cell.row][cell.col] = cell;
+                    });
+
+                    // Convert each row into a record with keys
+                    var header_row = rows[1];
+                    header_row = _.map(header_row, function(item, i) {
+                        if (item) {
+                            return item.value.toLowerCase();
+                        } else {
+                            return i;
+                        }
+                    });
+                    var record_rows = [];
+                    rows.slice(2).forEach(function(row) {
+                        record_rows.push(_.object(_.zip(header_row, row)));
+                    });
+                    callback(undefined, record_rows);
                 }
-                var row = potential_rows[0][0];
-                var reversed = potential_rows[0][1];
-                callback(undefined, row, reversed);
-            });
+            );
         });
     });
+}
 
+// Finds the given pairing in one of the team spreadsheets
+// callback gets three values: error, row, whether the pairing is reversed or not.
+function find_pairing(service_account_auth, spreadsheet_key, white, black, callback) {
+    var options = {
+        'min-col': 1,
+        'max-col': 6
+    };
+    function row_matches_pairing(row) {
+        var row_black = row.black.value.toLowerCase();
+        var row_white = row.white.value.toLowerCase();
+        if (
+            (row_black.indexOf(black) != -1 && row_white.indexOf(white) != -1)
+        ) {
+            return true;
+        } else if (
+            (row_white.indexOf(black) != -1 && row_black.indexOf(white) != -1)
+        ) {
+            return true;
+        }
+        return false;
+    }
+    get_rows(
+        service_account_auth,
+        spreadsheet_key,
+        options,
+        function(err, rows) {
```

Comment:
I wonder if it makes sense to diverge from the interface defined by the looks API and include a filter functor as a parameter. Then it might be easier to extend this to some of our other applications.